CabSignal Tweaks

icklesteve opened the pull request:
Use standard JMRI table button listener
TableModel to bundle
Unused imports
Added screenshot

icklesteve commented:
Failed HEADLESS=false STATIC=false  47 min 30 sec

```
[ERROR] Failures: 
[ERROR]   LinkedWarrantTest.testLinkedMidScript:340 "Tinker finished first leg" did not occur in time
[ERROR]   OptionFrameTest>OperationsTestCase.tearDown:63 Unexpected ERROR or higher messages emitted: null
[ERROR]   Z21SimulatorAdapterTest.tearDown:183 Level mismatch when looking for WARN message: "Timeout on reply to message:" found "Giving up after 3 attempts.  Exception binding to port 21105"
[INFO] 
[ERROR] Tests run: 40365, Failures: 3, Errors: 0, Skipped: 757
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 44:01 min
```

icklesteve commented:
Failed HEADLESS=false STATIC=false 25 min 46 sec

```
[INFO] Running jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

rhwood commented:
Can we please change the on/off icon to not use the track power button for this? I think it best that the track power button be used *only* to indicate track power status, not any on/off status in the application UI.

icklesteve commented:
Thanks for pointing out @rhwood , makes sense to keep that icon specifically for track power.
I've updated the icon used to something else and updated screenshot, if you could check that looks ok?

rhwood commented:
It looks okay, although, now that I think about it, since cab signaling (unless I’m missing something) is either on/off, it might be better to use a checkbox instead of a button with icon to show the on/off state (the power buttons are tri-state (on/off/unknown) which makes using a checkbox difficult).

icklesteve commented:
The original objective was a mechanism to send emergency CabSignal not valid instructions to all active cabs, hence I went for a big button for a user to aim for.
I'd say it's a button as it performs this action, extra to actually enabling / disabling something?

Individual CabSignals do have a standard 2-state on / off checkbox for each table row. ( which also performs the action for an individual CabSignal )

One of my personal rules is that each Pane I devise has at least 1 colour element, hence the button icon ;-)
Perhaps in future use the actual signal aspect graphic in the table? . . . .

rhwood commented:
The purpose of that button was not clear to me—let me think about it.

pabender commented:
I'm not entirely certain that button is necessary.

What it does right now is enumerates the individual cab signals and disables them or enables them.  

After pressing the button twice in a row, the old state of the individual rows is not restored.  If it really is a master switch, I would expect the original state to be restored.

In other words, if you were to click the checkbox on one row to turn that cab signal off, pressing the button once turns all the cab signals off and pressing it twice turns them all on, so we loose memory of the individual rows.

@icklesteve introduced the button in the original tool, but I can't remember if the turn all off/turn all on  behavior is something I introduced when turning this into a generic tool or something @icklesteve introduced in the MERG specific tool.

pabender commented on java/src/jmri/jmrit/cabsignals/CabSignalPane.java:
```diff
@@ -47,10 +39,11 @@
- * Created with Notepad++
+ * Pane for sending Cab Signal data via block lookup
  * @author Steve Young Copyright (C) 2018
- * @since 4.13.4
```
I would go ahead and leave these at 4.13.4.  That is when the original functionality was introduced, even though the generic tool came later.

icklesteve replied:
Thanks, makes sense.

rhwood commented:
I'd not place a button icon unless that icon has specific meaning in its context. In this case, a green light does not indicate "cab signals should be invalid", so I would remove the icon. (Especially if an operating instruction for an invalid cab signal is stop at next signal and call the dispatcher to get manual authority to proceed.)

I'd really avoid adding colors for the sake of adding colors--although it can help to signify something, using when it doesn't signify something will cause accessibility issues for visually impaired or those relying on the icon being reused to indicate something else (I'd assume green == good/go/on, so using green to invalidate signals does not match my instinctive interpretation of that color).

icklesteve commented:
Thinking further on this, I was mis-describing on / off, rather than what I intended as pause / resume.

There's no issue with disabling Cab Signals in urgent need, it's the re-enabling which is potentially more awkward.
On session resume following a system-wide signal outage, trying to remember which individual rows were checked prior to the outage was impossible.

Altered PR to restore original toggle functionality of pause / resume button not affecting individual tick status of rows.
Was not immediately apparent that this functionality had changed as the GUI was not being updated, 
although the correct CagSignal code was being sent to the layout.

Icon removed from pause / resume button, difference between Pause / Resume seems easier to tell from Off / On.
With the default text size, it's still awkward to read button text from a few meters away from the screen.
A JMRI-wide icon package, black / grey and white fine for these sort of icons would be great, but that's a different discussion.

ToolTips added for Pause / Resume button.
Grouped add loco buttons to distance function of pause / resume button.
Single CabSig active checkbox column moved to left of table.
Fixed issue with End Bumpers not being discovered if they weren't at the end of the start of the block path.
Added aspect icon column with menu item to set rotation if not correct orientation.
Updated documentation

bobjacobsen commented:
@pabender is this ready to go?

icklesteve commented on java/src/jmri/jmrit/cabsignals/CabSignalPane.java:
```diff
@@ -161,28 +153,32 @@ public void actionPerformed(ActionEvent e) {
+                        new NamedIcon("resources/icons/panels/CSD/AZD/button/button-green-off.GIF", 
+                        "resources/icons/panels/CSD/AZD/button/button-green-off.GIF"));
                     slotModel.masterSendCabData = false;
                     slotModel.masterSendCabDataButton(false);
```
modified as per comment below

icklesteve commented on java/src/jmri/jmrit/cabsignals/CabSignalPane.java:
```diff
@@ -161,28 +153,32 @@ public void actionPerformed(ActionEvent e) {
+                        new NamedIcon("resources/icons/panels/CSD/AZD/button/button-green.GIF",
+                        "resources/icons/panels/CSD/AZD/button/button-green.GIF"));
                     slotModel.masterSendCabData = true;
                     slotModel.masterSendCabDataButton(true);
```
modified as per changes to button functionality
